### PR TITLE
(feat) Added Bitflip Channel

### LINKED ISSUE #989: Add functions to generate quantum channels
Some channels are already defined in `toqito/channels`. We want to add a few more:

- [x] Amplitude damping channel
- [ ] Bit flip channel or Phase flip channel or Bit and Phase flip channel
- [x] Phase damping channel

### PR DESCRIPTION
**Pull request by RajdeepAher**
## Description

Related to #989 (partially)

Implemented the `bitflip` function to model quantum bit-flip . This function simulates a bit-flip operation occurring with a given probability.

## Changes

  -  [ ] Added `bitflip` function to simulate bit-flip.
  -  [ ] Implemented unit tests covering edge cases (probabilities 0, 1, and intermediate values).
  -  [ ] Added input validation to check probability values are within the range \([0,1]\).

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [ ] Use `ruff` for errors related to code style and formatting.
  -  [ ] Verify all previous and newly added unit tests pass in `pytest`.
  -  [ ] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR.
  -  [ ] Use `linkcheck` to check for broken links in the documentation.
  -  [ ] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

File: toqito/channels/bitflip.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    else:
+        raise ValueError("Bitflip channel is only defined for qubits (dim=2).")
+    kraus_ops = [k0, k1]
+
+    # If no input matrix is provided, return the Kraus operators
+    if input_mat is None:
+        return kraus_ops
+
+    # Apply the channel to the input state
+    # print(type(input_mat))
```
Comment (vprusso) [inline toqito/channels/bitflip.py:98]: Looks like an extra print statement was left here.

File: toqito/channels/bitflip.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    no_flip_prob = np.sqrt(1 - prob)
+    flip_prob = np.sqrt(prob)
+
+    # Identity matrix for the no-flip case
+    k0 = no_flip_prob * np.eye(dim)
+
+    # X gate for the flip case
+    if dim == 2:
+        k1 = flip_prob * np.array([[0, 1], [1, 0]])
+    else:
```
Comment (vprusso) [inline toqito/channels/bitflip.py:89]: If the bitflip channel is only defined when `dim == 2`, why do we make this a parameter in the function?

File: toqito/channels/bitflip.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        :filter: docname in docnames
+
+    :param input_mat: A matrix or state to apply the channel to. If None, returns the Kraus operators.
+    :param prob: The probability of a bitflip occurring.
+    :param dim: The dimension of the system (default is 2 for qubits).
+    :return: Either the Kraus operators of the bitflip channel if input_mat is None,
+             or the result of applying the channel to input_mat.
+
+    """
+    if prob < 0 or prob > 1:
```
Comment (vprusso) [inline toqito/channels/bitflip.py:68]: You can write this more succinctly as `if not (0 <= prob <= 1):`

File: toqito/channels/bitflip.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        raise ValueError("Probability must be between 0 and 1.")
+
+    # If input is a CVX variable, handle it appropriately
+    if isinstance(input_mat, Variable):
```
Comment (vprusso) [inline toqito/channels/bitflip.py:72]: I'm not sure why we would need to cover the case here when the `input_mat` is a CVXPY-type object.

File: toqito/channels/tests/test_bitflip.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+"""Test cases for the bit-flip channel in Toqito."""
```
Comment (vprusso) [inline toqito/channels/tests/test_bitflip.py:1]: We can remove the "in Toqito" part of the this docstring.

File: toqito/channels/tests/test_bitflip.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+"""Test cases for the bit-flip channel in Toqito."""
+
+import unittest
```
Comment (vprusso) [inline toqito/channels/tests/test_bitflip.py:3]: We should use `pytest` (not `unittest`) to conform with how testing is done elsewhere in `toqito` for consistency.

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -1,10 +1,10 @@
...
-- repo: https://github.com/astral-sh/ruff-pre-commit
-  # Ruff version (keep in sync with the version in pyproject.toml)
-  rev: v0.8.2
-  hooks:
-    # Run the linter.
-    - id: ruff
-      args: [ --fix ]
-    # Run the formatter.
-    - id: ruff-format 
+repos:
```
Comment (vprusso) [inline .pre-commit-config.yaml:1]: Seems okay, but since these seem identical, I wonder why GitHub is flagging them as having changed. Can probably ignore this comment, but mentioning it as it also seems to be the case in `channels/__init__.py` as well.

**Review by vprusso — COMMENTED** [review]
Thanks for your contribution, @RajdeepAher ! 

I've added a few comments here on the MR.

File: toqito/channels/bitflip.py (hunk 1 above)
Comment (RajdeepAher) [reply toqito/channels/bitflip.py:98]: Yes, will remove it. Thanks

File: toqito/channels/bitflip.py (hunk 2 above)
Comment (RajdeepAher) [reply toqito/channels/bitflip.py:89]: I felt that my understanding was somewhat lacking, and there might be a future extension of this to higher dimensions that I am not aware of at the moment and hence the parameter. I can remove it too!

File: toqito/channels/bitflip.py (hunk 4 above)
Comment (RajdeepAher) [reply toqito/channels/bitflip.py:72]: I saw rest of the implementations in `toqito/channels` and thought it would be better to include CVXPY object. I can remove it , if it is wrong

File: .pre-commit-config.yaml (hunk 7 above)
Comment (RajdeepAher) [reply .pre-commit-config.yaml:1]: yes, even I am not aware of why it is happening

**Comment by purva-thakre** [comment]
@RajdeepAher Please make sure your function is properly covered by unit tests. 

image

**Comment by RajdeepAher** [comment]
> @RajdeepAher Please make sure your function is properly covered by unit tests.
> 
> image

Hello,
The error in coverage was due to inclusion of CVXPY-type object, but now I have removed it

**Comment by purva-thakre** [comment]
? There are a bunch of test failures though..

https://github.com/vprusso/toqito/actions/runs/13677568697/job/38245804389?pr=990

https://github.com/vprusso/toqito/actions/runs/13677568689/job/38245803706?pr=990

File: toqito/channels/bitflip.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+def bitflip(
+    input_mat: np.ndarray | None,
+    prob: float = 0,
+) -> np.ndarray:
+    r"""Apply the bitflip quantum channel to a state or return the Kraus operators.
+
+    The *bitflip channel* is a quantum channel that flips a qubit from :math:`|0\rangle` to :math:`|1\rangle`
+    and from :math:`|1\rangle` to :math:`|0\rangle` with probability :math:`p`. It is defined by
+    the following operation:
+    .. math::
```
Comment (purva-thakre) [inline toqito/channels/bitflip.py:17]: Add a blankline before math mode. 

https://toqito--990.org.readthedocs.build/en/990/autoapi/channels/bitflip/index.html

File: toqito/channels/bitflip.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    >>> rho = np.array([[1, 0], [0, 0]])  # |0><0|
+    >>> bitflip(rho, prob=0.3)
+    array([[0.7, 0. ],
+           [0. , 0.3]])
+
+    References
+    ==========
+    .. bibliography::
+        [1] https://arxiv.org/abs/1106.1445
+        [2] https://www.quantumchannelzoo.org/person/bit-flip
```
Comment (purva-thakre) [inline toqito/channels/bitflip.py:57]: Add these to `refs.bib` instead of directly listing links to references. 

https://toqito.readthedocs.io/en/latest/contributing.html#references-in-docstrings

File: toqito/channels/tests/test_bitflip.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+from toqito.channels import bitflip
+
+
+def test_kraus_operators():
+    """Test if the function returns correct Kraus operators for given probability."""
+    prob = 0.3
+    kraus_ops = bitflip(prob=prob)
+    expected_k0 = np.sqrt(1 - prob) * np.eye(2)
+    expected_k1 = np.sqrt(prob) * np.array([[0, 1], [1, 0]])
```
Comment (purva-thakre) [inline toqito/channels/tests/test_bitflip.py:14]: You can create a list of these two and directly compare the calculated output to the expected output.

File: toqito/channels/tests/test_bitflip.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    result = bitflip(rho, prob=0)
+    np.testing.assert_almost_equal(result, rho)
+
+
+def test_probability_1():
+    """Test bitflip with probability 1 (always flips)."""
+    rho = np.array([[0.5, 0.5], [0.5, 0.5]])
+    expected_output = np.array([[0.5, 0.5], [0.5, 0.5]])  # Same for maximally mixed state
+    result = bitflip(rho, prob=1)
+    np.testing.assert_almost_equal(result, expected_output)
```
Comment (purva-thakre) [inline toqito/channels/tests/test_bitflip.py:52]: Same as my other comment, most of these unit tests can be condensed into 1 by using using parameterize.

File: toqito/channels/bitflip.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+    :param input_mat: A matrix or state to apply the channel to. If None, returns the Kraus operators.
+    :param prob: The probability of a bitflip occurring.
+    :return: Either the Kraus operators of the bitflip channel if input_mat is None,
+             or the result of applying the channel to input_mat.
```
Comment (purva-thakre) [inline toqito/channels/bitflip.py:62]: If you are referring to a parameter, sandwich it between `

File: toqito/channels/bitflip.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        raise ValueError("Probability must be between 0 and 1.")
+
+    if input_mat is not None:
+        dim = input_mat.shape[0]
+    else:
+        dim = 2
+
+    # Enforce 2-dimensional (qubit) requirement
+    if dim != 2:
+        raise ValueError("Bitflip channel is only defined for qubits (dim=2).")
```
Comment (purva-thakre) [inline toqito/channels/bitflip.py:75]:
```suggestion
    if input_mat is not None:
        if input_mat.shape[0] != 2:
              raise ValueError("Bitflip channel is only defined for qubits (dim=2).")
    
      
   dim = 2

```

File: docs/refs.bib [hunk 14]
```diff
@@ -1,1454 +1,1462 @@
-# File for all article, arxiv and other web references
```
Comment (purva-thakre) [inline docs/refs.bib:1]: why were the items in this file deleted and added back in ?

Comment (RajdeepAher) [reply docs/refs.bib:1]: Sorry, I had actually only added the last few lines but it somehow modified the entire file. Working to fix it

File: toqito/channels/bitflip.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
+"""Implements the bitflip quantum gate channel."""
+
+import numpy as np
+
+from toqito.helper import expr_as_np_array, np_array_as_expr
```
Comment (vprusso) [inline toqito/channels/bitflip.py:5]: These are imported but unused.

File: toqito/channels/bitflip.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+            raise ValueError("Bitflip channel is only defined for qubits (dim=2).")
+
+    dim = 2
+
+    # Define the Kraus operators for the bitflip channel
+    no_flip_prob = np.sqrt(1 - prob)
+    flip_prob = np.sqrt(prob)
+
+    # Identity matrix for the no-flip case
+    k0 = no_flip_prob * np.eye(dim)
```
Comment (vprusso) [inline toqito/channels/bitflip.py:82]: We can likely just remove `dim` and put `2` here to get rid of the `dim = 2` variable declaration.

File: toqito/channels/bitflip.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+    # X gate for the flip case
+    k1 = flip_prob * np.array([[0, 1], [1, 0]])
+
+    kraus_ops = [k0, k1]
+
+    # If no input matrix is provided, return the Kraus operators
```
Comment (vprusso) [inline toqito/channels/bitflip.py:89]: [nit]: Period at end of comments.

File: toqito/channels/bitflip.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    if input_mat is None:
+        return kraus_ops
+
+    # Apply the channel to the input state
+    input_mat = np.asarray(input_mat, dtype=complex)
+
+    result = np.zeros_like(input_mat, dtype=complex)
+    for op in kraus_ops:
+        result += op @ input_mat @ op.conj().T
+    return result
```
Comment (vprusso) [inline toqito/channels/bitflip.py:99]: Could we not just do:

```suggestion
    return k0 @ input_mat @ k0.conj().T + k1 @ input_mat @ k1.conj().T
```

File: toqito/channels/bitflip.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    no_flip_prob = np.sqrt(1 - prob)
+    flip_prob = np.sqrt(prob)
+
+    # Identity matrix for the no-flip case
+    k0 = no_flip_prob * np.eye(dim)
+
+    # X gate for the flip case
+    k1 = flip_prob * np.array([[0, 1], [1, 0]])
+
+    kraus_ops = [k0, k1]
```
Comment (vprusso) [inline toqito/channels/bitflip.py:87]:
```suggestion

```

File: toqito/channels/bitflip.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+
+    # If no input matrix is provided, return the Kraus operators
+    if input_mat is None:
+        return kraus_ops
```
Comment (vprusso) [inline toqito/channels/bitflip.py:91]:
```suggestion
        return [k0, k1]

```

File: toqito/channels/bitflip.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+    if input_mat is not None:
+        if input_mat.shape[0] != 2:
+            raise ValueError("Bitflip channel is only defined for qubits (dim=2).")
+
+    dim = 2
+
+    # Define the Kraus operators for the bitflip channel
+    no_flip_prob = np.sqrt(1 - prob)
+    flip_prob = np.sqrt(prob)
```
Comment (vprusso) [inline toqito/channels/bitflip.py:79]:
```suggestion
```

File: toqito/channels/bitflip.py (hunk 16 above)
Comment (vprusso) [inline toqito/channels/bitflip.py:82]:
```suggestion
    k0 = np.sqrt(1 - prob) * np.eye(2)
```

File: toqito/channels/bitflip.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+    # Identity matrix for the no-flip case
+    k0 = no_flip_prob * np.eye(dim)
+
+    # X gate for the flip case
+    k1 = flip_prob * np.array([[0, 1], [1, 0]])
```
Comment (vprusso) [inline toqito/channels/bitflip.py:85]:
```suggestion
    k1 = np.sqrt(prob) * np.array([[0, 1], [1, 0]])
```

File: toqito/channels/bitflip.py (hunk 21 above)
Comment (RajdeepAher) [reply toqito/channels/bitflip.py:79]: I tried to keep code more intuitive to understand and readable at first try, will condense it

File: toqito/channels/bitflip.py [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    :return: Either the Kraus operators of the bitflip channel if `input_mat` is `None`,
+             or the result of applying the channel to `input_mat`.
+
+    """
+    if not (0 <= prob <= 1):
+        raise ValueError("Probability must be between 0 and 1.")
+
+    if input_mat is not None:
+        if input_mat.shape[0] != 2:
+            raise ValueError("Bitflip channel is only defined for qubits (dim=2).")
```
Comment (vprusso) [inline toqito/channels/bitflip.py:71]:
```suggestion
    if input_mat is not None:
        if input_mat.shape[0] != 2:
            raise ValueError("Bitflip channel is only defined for qubits (dim=2).")
    else:
        # If no input matrix is provided, return the Kraus operators.
        return [k0, k1]
```

File: toqito/channels/bitflip.py [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+
+    # Identity matrix for the no-flip case.
+    k0 = np.sqrt(1 - prob) * np.eye(2)
+
+    # X gate for the flip case.
+    k1 = np.sqrt(prob) * np.array([[0, 1], [1, 0]])
+
+    # If no input matrix is provided, return the Kraus operators.
+    if input_mat is None:
+        return [k0, k1]
```
Comment (vprusso) [inline toqito/channels/bitflip.py:81]:
```suggestion
```

File: toqito/channels/bitflip.py [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+    :param prob: The probability of a bitflip occurring.
+    :return: Either the Kraus operators of the bitflip channel if `input_mat` is `None`,
+             or the result of applying the channel to `input_mat`.
+
+    """
+    if not (0 <= prob <= 1):
+        raise ValueError("Probability must be between 0 and 1.")
+
+    if input_mat is not None:
+        if input_mat.shape[0] != 2:
```
Comment (vprusso) [inline toqito/channels/bitflip.py:70]: What if someone sends in a 2 x N matrix?

File: toqito/channels/tests/test_bitflip.py [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+def test_apply_to_state(rho, expected_output, prob):
+    """Test bitflip application to |0><0| and |1><1| states."""
+    result = bitflip(rho, prob=prob)
+    np.testing.assert_almost_equal(result, expected_output)
+
+
+@pytest.mark.parametrize(
+    "rho, prob, expected_output",
+    [
+        # Test probability 0 (no change)
```
Comment (vprusso) [inline toqito/channels/tests/test_bitflip.py:37]:
```suggestion
        # Test probability 0 (no change).
```

File: toqito/channels/tests/test_bitflip.py [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+        (np.array([[0.5, 0.5], [0.5, 0.5]]), 0, np.array([[0.5, 0.5], [0.5, 0.5]])),
+        # Test probability 1 (always flips, but same for max mixed state)
```
Comment (vprusso) [inline toqito/channels/tests/test_bitflip.py:39]:
```suggestion
        # Test probability 1 (always flips, but same for max mixed state).
```

File: toqito/channels/tests/test_bitflip.py [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+def test_invalid_probability(prob, error_message):
+    """Test that invalid probabilities raise an error."""
+    with pytest.raises(ValueError, match=error_message):
+        bitflip(prob=prob)
+
+
+def test_invalid_dimension():
+    """Test that invalid dimensions raise an error."""
+    rho = np.eye(3)  # 3x3 matrix
+    with pytest.raises(ValueError, match="Bitflip channel is only defined for qubits"):
```
Comment (vprusso) [inline toqito/channels/tests/test_bitflip.py:65]:
```suggestion
    with pytest.raises(ValueError, match="Bitflip channel is only defined for qubits."):
```

File: toqito/channels/bitflip.py (hunk 23 above)
Comment (RajdeepAher) [reply toqito/channels/bitflip.py:71]: Hello,

Writing an else statement here would be redundant since the function already returns if input_mat is None. Instead, I suggest structuring it like this:

```
if input_mat is None:
    return [k0, k1]
if input_mat.shape != (2, 2):
    raise ValueError("Input matrix must be 2x2 for the bitflip channel.")
```    
Let me know what you think.

Comment (vprusso) [reply toqito/channels/bitflip.py]: Yes, that should work!

**Review by vprusso — APPROVED** [review]
I'll approve but will wait for @purva-thakre to provide her review as well. Nice work, @RajdeepAher !
